### DISCUSSION TITLE
Correct way to convert a ModbusMessage to a String

### DISCUSSION THREAD

**Discussion by baggior** [eModbus/eModbus#91] [eModbus applications]
For log purpose I need to convert a **ModbusMessage to a String**, what is the suggested way?

I'm also wondering why there is not a **toString()** public method inside the ModbusMessage  class!

**Comment by Miq1** [comment]
Well, what is the String supposed to look like? Most messages bytes are in the non-printable range, so the closest thing I can offer is to use the ``HEXDUMP_N()`` function from ``Logging.h``. It was in fact created to log out Modbus data.

See [the Logging docs](https://emodbus.github.io/logging#logging-statements) for application.

**Reply by baggior** [reply]
i've managed to use this
` HEXDUMP_N("INPUT MESSAGE BODY", request.data(), request.size() );`

but is not what I need beacause it directly outputs to the Serial Print, instead I would rather prefer to have a String I could elaborate further..

So I would suggest a method like:

```
String ModbusMessage::toString() {
   String ret;
   //TODO populate the ret String with the hex dump of the message body, using a method similar to "logHexDump()"

   return ret;
}
```

**Reply by Miq1** [reply]
Why don't you just go ahead and write it? Should not be too difficult - use a range ``for`` loop on the ``ModbusMessage`` and add a hexadecimal value to your string for each byte.

**Reply by Miq1** [reply]
Another alternative: use a ``StreamString`` type object, and assign that to ``LOGDEVICE``.  That should be possible, as it is derived from ``Stream`` that in turn is a derivation of ``Print``.

Then all ``Logging`` output will be directed into that object, that is usable as a ``String`` as well. DId not try that, but something along of
```
#include <StreamString.h>
#include "Logging.h"

StreamString sString;

void setup() {
  LOGDEVICE = &sStream;
...

  HEXDUMP_N("Foo", bar.data(), bar.size()); 
  Serial.prinf(sStream);
...
```

**Reply by baggior** [reply]
I know I could do many things even outside the class... but the original question was..
why not include this simple and useful code just directly inside the `ModbusMessage `class implementation ??

**Reply by Miq1** [reply]
Because there are other ways, as I suggested, and there is no urge to have it?
